### ISSUE TITLE
IsAuthenticated Permission on LoginView

### ISSUE THREAD

**Issue by cwkang1998** [James1345/django-rest-knox#98]
Forgive me if I misunderstand the implementation for the LoginView for knox, but shouldn't the LoginView have a permission of 'Allow Any' instead of 'IsAuthenticated' ?

**Comment by belugame** [comment]
No worries, that behavior indeed looks illogical on first read as compared to Django's.
Maybe we should reconsider the naming...

Think of LoginView as CheckCredentialsAndReturnATokenView. You need to send username and password via e.g. Basic-Auth headers to that view and it will return a token to you if they match an existing user. 

Now if you had AllowAny, it would give everyone a token without having to provide any sort of credentials, which makes authentication redundant.

**Comment by cwkang1998** [comment]
@belugame 
Thank you for your reply, a few further questions here and some confirmation:

> Think of LoginView as CheckCredentialsAndReturnATokenView. You need to send username and password via e.g. Basic-Auth headers to that view and it will return a token to you if they match an existing user.

Since I am using KnoxTokenAuth as the default authentication, am I supposed to change it to BasicAuth from django-rest for authentication specifically for this endpoint and explicitly state that other endpoints will be using KnoxTokenAuth? If so, it is not stated in the documentation (but it is stated in Django Rest Authentication Page) so its a little bit confusing.

I am currently implementing my own LoginView by extending the KnoxLoginView to login using credentials directly (providing user and password and receiving a token without first going through a BasicAuthentication) while still setting KnoxTokenAuth as my default authentication.

**Comment by cwkang1998** [comment]
For your reference, here is my custom implementation:

`from django.contrib.auth import login

from rest_framework import permissions
from rest_framework.authtoken.serializers import AuthTokenSerializer
from knox.views import LoginView as KnoxLoginView

class LoginView(KnoxLoginView):
    permission_classes = (permissions.AllowAny,)

    def post(self, request, format=None):
        serializer = AuthTokenSerializer(data=request.data)
        serializer.is_valid(raise_exception=True)
        user = serializer.validated_data['user']
        login(request, user)
        return super(LoginView, self).post(request, format=None)`

**Comment by belugame** [comment]
I'm not sure what you are trying to achieve... what do you send to your LoginView? A token? I would expect username/password.

> providing user and password and receiving a token without first going through a BasicAuthentication

Providing username/password and getting a token is what you can do with BasicAuthentication. Maybe you were thinking that is an extra step.
This is how you would get a token via basic auth using httpie (check it out, it's a great helper). 

`http -jva user@host.com:mypassword POST host.com/api/login/`

**Comment by cwkang1998** [comment]
> I'm not sure what you are trying to achieve... what do you send to your LoginView? A token? I would expect username/password.
Yes, i do send my username and password to my login view

**Comment by cwkang1998** [comment]
Ok I played around a little bit and this is what I found out:

1) With no default authentication classes specified in DRF, basic authentication will be used. Thus, LoginView will work without any issues.
2) However, with my setting in which KnoxTokenAuthentication is the only authentication class, I would need to create a custom view to accept credentials as the default implementated LoginView for Knox uses the default authentication classes for the view. Thus, when I attempt to go to the LoginView with such settings, it requires a Token to authenticate, but I do not have one yet at that point, thus my problem.
3) So, for me to solve the issues, either I can create a custom view to keep my settings, or I can add one more authentication scheme in the default authentication class, and explicitly state which views will use the KnoxTokenAuthentication scheme.

Thank you very much for your patience.
